juju/application.py

```diff
@@ -706,19 +701,23 @@ async def refresh(
         if parsed_url.schema is None:
             raise JujuError(f'A ch: or cs: schema is required for application refresh, given : {str(parsed_url)}')
 
-        if revision is not None:
-            origin.revision = revision
+        current_origin = charm_url_origin_result.charm_origin
+        if path is not None or (switch is not None and is_local_charm(switch)):
+            await self.local_refresh(current_origin, force, force_series,
+                                     force_units, path or switch, resources)
+            return
 
-        # Make the source-specific changes to the origin/channel/url
-        # (and also get the resources necessary to deploy the (destination) charm -- for later)
-        origin.source = Source.CHARM_HUB.value
-        if channel:
-            ch = Channel.parse(channel).normalize()
-            origin.risk = ch.risk
-            origin.track = ch.track
+        if channel is not None:
+            channel = Channel.parse(channel).normalize()
 
-        charmhub = self.model.charmhub
-        charm_resources = await charmhub.list_resources(charm_name)
+        origin = client.CharmOrigin(
+            source=str(Source.CHARM_HUB),
+            risk=channel.risk if channel else current_origin.risk,
+            track=channel.track if channel else current_origin.track,
+            revision=revision or current_origin.revision,
+            base=current_origin.base,
+            architecture=current_origin.architecture or DEFAULT_ARCHITECTURE,
```

Comment:
I'm not sure if the architecture is gonna be set every time, so you might wanna 

```suggestion
            architecture=current_origin.get('architecture', DEFAULT_ARCHITECTURE),
```

Reply:
Fixed